### ISSUE TITLE
IOError when migrating to new computer

### ISSUE THREAD

**Issue by aliceriot** [sup-heliotrope/sup#442]
Hey folks,

I'm currently setting up a new laptop, and I want to move my sup setup to the new machine while keeping all my labels and so on.

This is the process I took:

- update sup on old machine
- kill getmail cronjob
- run `sup-sync`
- copy `~/mail` (location of maildirs) to flash drive
- copy `~/.sup` to flash drive
- install sup (w/ rubygems) on new machine
- ran it once to make sure it was working
- deleted initial (default) `.sup` (on new machine)
- copied `mail` and `.sup` to ~/ on new machine

And now the fun, I get the following exception:

    --- IOError from thread: main
    DatabaseError: Error reading block 5693: got end of file
    /home/alice/.gem/ruby/2.2.0/gems/sup-0.22.1/lib/sup/index.rb:120:in `initialize'
    /home/alice/.gem/ruby/2.2.0/gems/sup-0.22.1/lib/sup/index.rb:120:in `new'
    /home/alice/.gem/ruby/2.2.0/gems/sup-0.22.1/lib/sup/index.rb:120:in `load_index'
    /home/alice/.gem/ruby/2.2.0/gems/sup-0.22.1/lib/sup/index.rb:103:in `load'
    /home/alice/.gem/ruby/2.2.0/gems/sup-0.22.1/lib/sup/util.rb:610:in `method_missing'
    /home/alice/.gem/ruby/2.2.0/gems/sup-0.22.1/bin/sup:144:in `<module:Redwood>'
    /home/alice/.gem/ruby/2.2.0/gems/sup-0.22.1/bin/sup:74:in `<top (required)>'
    /home/alice/.gem/ruby/2.2.0/bin/sup:23:in `load'
    /home/alice/.gem/ruby/2.2.0/bin/sup:23:in `<main>'

Anyone have an idea what's up? The systems are both updated Arch Linux systems, both running sup v0.22.1 and ruby 2.2.2p95. I'd really like to avoid rebuilding the index...

Thanks!

**Comment by gauteh** [comment]
Hi,

looks like an corrupt xapian index. Don't know what happened, you could
try to copy it again.

What could also work is to use a combination of sup-dump and
sup-import-dump, but you probably need to rebuild your index first
anyway.

You could also try to use an older version of xapian. In that case it is
an xapian issue.

**Comment by aliceriot** [comment]
Hmm, ok I'll give that a try when I get home tonight (old machine is at home). The strange thing is that the index as it is now works fine on the old machine - I can open sup, perform searches, etc. It only causes problems when I transfer it to the new machine. I think I'll try doing sup-dump on the old machine, transferring the dump, and using dump-import on the new one. Stay tuned!

**Comment by aliceriot** [comment]
Also I just realized I didn't check to see if the same version of Xapian was installed on the two machines, that could be another source of the mismatch.

**Comment by rakoo** [comment]
Did you close sup cleanly on the old machine ? If old sup was in the middle of an index modification then maybe it didn't write everything cleanly. Closing sup makes sure the index is in a correct state.

**Comment by aliceriot** [comment]
Yes, sup was closed. I ran sup-sync separately, thinking that it would be more likely to leave the index in a clean state. I suspect the problem is different versions of Xapian.

**Comment by aliceriot** [comment]
Sorted it out! Turns out what worked was:

- do `sup-dump` on old machine
- delete `xapian` in `.sup` on new machine (`.sup` is otherwise identical to old machine)
- run `sup-sync` on new machine to add messages to index (I had already copied over my maildir)
- run `sup-import-dump` to get message state from the dump done earlier

And it works! huzzah! Thanks for the help everyone!

Side note: building the index takes a reeeally long time when you have ~50k emails.
